### ISSUE TITLE
Add override for sem_open()

### ISSUE THREAD

**Issue by mardy** [sergiusens/snapcraft-preload#2]
When `sem_open()` is given the `O_CREAT` or `O_EXCL` flag, glibc first creates a temporary file `/dev/shm/<random>` and only after the creation is successful it calls `link()` and links it to the filename chosen by the client. See the thread at https://lists.ubuntu.com/archives/snapcraft/2017-February/003120.html

I think we should intercept the `sem_open()` call and -- if `O_CREAT` or `O_EXCL` are given -- do this instead, as suggested by Jamie:

```
   char *name = "/dev/shm/sem.snap.foo.bar";
   open(name, O_CREAT | O_EXCL | O_RDRW, S_IRUSR | S_IWUSR);
   sem_open("snap.foo.bar", 0);
```

I would also suggest that we go a bit further and always ensure that the name starts with `"snap.$SNAP_NAME"`.

**Comment by mardy** [comment]
I can take care of this, if you like. Just let me know :-)

**Comment by sergiusens** [comment]
Yes please. I only put this up to get something started but I haven't done C in a long while.

**Comment by diddledan** [comment]
I appear to be hitting the problem that this issue is raised to discuss, namely a snap that I'm working on (being closed source) is trying to open a semaphore which glibc is rewriting to /dev/shm/<random> and thus failing.

@mardy, @sergiusens: Have either of you worked on the fix at all yet? if so what's the current status on getting the suggested workaround implemented here?

**Comment by mardy** [comment]
@diddledan I started working on it here (https://github.com/mardy/snapcraft-preload/tree/sem_open) but I don't remember what's the state is. For sure it's not ready, since I didn't propose it; I seem to recall that the problem I met was that there was some function in libc that I would have to completely re-implement in the preloader, at which point I lost interest in it.

**Comment by dhoomakethu** [comment]
@mardy I followed your thread on the ubuntu archives and I am right now in the exact same problem as you were (python multiprocessing + /dev/shm + strict confinement) . Would you mind sharing any hints on how you actually overcame that issue ? I tried the latest of snapcraft-preload but it doen't seem to be fixing. Thanks in advance.

**Comment by mardy** [comment]
Hi @dhoomakethu ! Sorry, I stopped working on this stuff, so I have no solution. :-(

**Comment by izznogooood** [comment]
It seems I have hit this problem to:

```
  File "/snap/ftpscan/x1/usr/lib/python3.6/multiprocessing/synchronize.py", line 162, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/snap/ftpscan/x1/usr/lib/python3.6/multiprocessing/synchronize.py", line 59, in __init__
    unlink_now)
PermissionError: [Errno 13] Permission denied
```

Has anyone found a workaround?

My snap is a net"worker" and pretty pointless without multitasking :)

**Comment by artivis** [comment]
I have hit this issue too (python multiprocessing + /dev/shm + strict confinement) while trying to snap a Python application that uses [Ansible](https://github.com/ansible/ansible) Python API.  

Still no fix/workaround?

**Comment by diddledan** [comment]
This is already fixed. Simply use `snapcraft-preload` and you should be set now.

**Comment by artivis** [comment]
Hi @diddledan, I gave `snapcraft-preload` a shot in the meantime and hit another issue.  
For some context, the Python app I'm snapping 'git clone' a repo before running some Ansible task out of it. *With* `snapcraft-preload`, the app fails running a `git version` command whereas it goes fine without it.  

The actual error is the following:
```terminal
/bin/sh: 1: cannot create /dev/null: Read-only file system
Traceback (most recent call last):
  File "**/lib/python3.6/site-packages/git/__init__.py", line 83, in <module>
    refresh()
  File "**/lib/python3.6/site-packages/git/__init__.py", line 73, in refresh
    if not Git.refresh(path=path):
  File "**/lib/python3.6/site-packages/git/cmd.py", line 215, in refresh
    cls().version()
  File "**/lib/python3.6/site-packages/git/cmd.py", line 542, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "**/lib/python3.6/site-packages/git/cmd.py", line 1005, in _call_process
    return self.execute(call, **exec_kwargs)
  File "**/lib/python3.6/site-packages/git/cmd.py", line 822, in execute
    raise GitCommandError(command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git version
  stderr: 'fatal: open /dev/null or dup failed: Read-only file system'
```

Running `git version` in a shell to my app works fine tho.
Any clue as what's going on and how to fix it? Should I open a new ticket?

Note that I still don't know if the initial issue is fixed.
